### ISSUE TITLE
【AIコース】大課題 タイタニック生存予測 ランダムフォレスト

### ISSUE THREAD

**Issue by hayyoshi** [hayyoshi/DIC#2]
### データ取得
# https://www.kaggle.com/c/titanic/data
### データ読み込み
import pandas as pd
df = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')
print(df.sample(10))
### 前処理について記述せよ
# 与えられたデータを加工する事を前処理と言う。前処理を行なう理由として、
# 例えば、特徴量に欠損値が多く含まれている状態や、型が統一されていない状態の場合、学習を行なっとしても
# 期待する精度が出ない、想定外の結果が発生してしまう為、それらの問題を未然に防ぐといった事が挙げられる。
# 前処理は、具体的には、欠損値を修正、データの型の統一、外れ値の正規化といったものになる。
# これにより、欠損値を修正する事で、選択した特長量を用いた学習を行う、外れ値の対応により、他のデータ、特徴量への影響を制御する、
# データの型の統一により、特徴量として使えるデータにする事が出切る。
### 2/前処理について記述せよ
# 前処理について記述せよの調査により、データを確認する際にどのような点を見るとよいかと言うと、以下の3つが挙げられる。
# 1.欠損値を修正
# 2.データの型の統一
# 3.外れ値の正規化
### 前処理を行う
# 前処理1
def simplify_ages(df):
    df.Age = df.Age.fillna(-0.5)
    bins = (-1, 0, 5, 12, 18, 25, 35, 60, 120)
    group_names = ['Unknown', 'Baby', 'Child', 'Teenager', 'Student', 'Young Adult', 'Adult', 'Senior']
    categories = pd.cut(df.Age, bins, labels=group_names)
    df.Age = categories
    return df

def simplify_cabins(df):
    df.Cabin = df.Cabin.fillna('N')
    df.Cabin = df.Cabin.apply(lambda x: x[0])
    return df

def simplify_fares(df):
    df.Fare = df.Fare.fillna(-0.5)
    bins = (-1, 0, 8, 15, 31, 1000)
    group_names = ['Unknown', '1_quartile', '2_quartile', '3_quartile', '4_quartile']
    categories = pd.cut(df.Fare, bins, labels=group_names)
    df.Fare = categories
    return df

def format_name(df):
    df['Lname'] = df.Name.apply(lambda x: x.split(' ')[0])
    df['NamePrefix'] = df.Name.apply(lambda x: x.split(' ')[1])
    return df    

def drop_features(df):
    return df.drop(['Ticket', 'Name', 'Embarked'], axis=1)

def transform_features(df):
    df = simplify_ages(df)
    df = simplify_cabins(df)
    df = simplify_fares(df)
    df = format_name(df)
    df = drop_features(df)
    return df

data_train = transform_features(df)
data_test = transform_features(df_test)
print(data_train.head())

# 前処理2
from sklearn import preprocessing
def encode_features(df_train, df_test):
    features = ['Fare', 'Cabin', 'Age', 'Sex', 'Lname', 'NamePrefix']
    df_combined = pd.concat([df_train[features], df_test[features]])

    for feature in features:
        le = preprocessing.LabelEncoder()
        le = le.fit(df_combined[feature])
        df_train[feature] = le.transform(df_train[feature])
        df_test[feature] = le.transform(df_test[feature])
    return df_train, df_test

data_train, data_test = encode_features(data_train, data_test)
data_train.head()

### 前処理の内容について記述せよ
#simplify_ages 欠損値を-0.5に置換し、年齢別でグループ分けしている。
#simplify_cabins　欠損値をNに置換し、最初の1文字目だけ抽出している。
#simplify_fares　欠損値を-0.5に置換し、運賃別でグループ分けしている。
#format_name　名字と接頭辞を抽出している。。
#drop_features　'Ticket', 'Name', 'Embarked'のカラムをdropしている。
#encode_features　トレーニング、テストデータに対し、特徴量として['Fare', 'Cabin', 'Age', 'Sex', 'Lname', 'NamePrefix']カラムを選択し、
#                結合処理を行なった上で、labelEncoderにより数値化し、リターンしている。
# 上記の処理を行なう事で、グループ分けによるータの傾向分析、欠損値の修正、外れ値の正規化、重要な特徴量のみを選択、ならびに文字を数値化し
# データの型の統一を行なっている。

### モデル選択について記述せよ
# 今回は、生存予測（分類）を行いますが、この分類について使用できそうな手法としては、以下の4つが考えられる。
#  1.ロジスティック回帰
#  2.単純ベイズ(ナイーブベイズ)
#  3.SVM
#  4.ランダムフォレスト
# 上記の手法の概要としては以下の通りです。
# 1.ロジスティック回帰:与えられた特徴量(説明変数)の相関関係をある関数(直線モデル)によって説明する。
# 2.単純ベイズ(ナイーブベイズ):ベイズの定理を活用し、特徴量(説明変数)に対して、最もその分類が発生する確率が高いものを予測する。
# 3.SVM:2つのクラスを分割するために、最も近い点との距離（マージン）を最大化するような超平面をとる。
# 4.ランダムフォレスト:アンサブル学習の一種で、複数の決定木を作り多数決で分類を行う。

# 以下にその手法の長所/短所をそれぞれ3つ挙げます。
# まず、今回の手法のうち、線形のモデルはロジスティック回帰、単純ベイズ(ナイーブベイズ)、SVMとなり、非線形のモデルはランダムフォレストとなります。
# 線形回帰と非線形回帰の基本的な違いは、モデルが取り得る関数の形式が挙げられます。
# その上で、線形の長所としては、以下の3つとなります。
# 　　-数式で出来る為、モデルの構造が分かり易い。
# 　　-数式で出来る為、予測が可能である。
# 　　-連続値のデータ分析を得意とする。
# 線形の短所としては、以下の3つとなります。
# 　　-非線形に比べてモデルの表現力が低い。
# 　　-特徴量を高次(2乗、3乗..)にした際に過学習を起こしやすい。
# 　　-離散型のデ－タ分析は不得意である
# 非線形の長所としては、以下の3つとなります。
# 　　-線形よりもモデルの表現力が高い。
# 　　-モデルの構造が分かり辛い。
# 　　-離散型のデータ分析を得意とする。
# 非線形の短所としては、以下の3つとなります。
# 　　-表現力が高い分、最適なモデルを決定するのに時間を要する。
# 　　-モデルの精度が良かった場合でも、なぜ良かったのか分かり辛い。
# 　　-表現力の高いモデルである分、線形と比べて精度を出す為のデータ量が多くなる。
# 以下、重複する箇所もありますが、各手法の長所となります。
# 1.ロジスティック回帰
# 長所
# 　　-数式で出来る為、モデルの構造が分かり易い。
# 　　-予測結果が0から1の間の数式を作れる。
# 　　-連続値のデータ分析を得意とする。
# 短所
# 　　-非線形に比べてモデルの表現力が低い。
# 　　-特徴量を高次(2乗、3乗..)にした際に過学習を起こしやすい。
# 　　-離散型のデ－タ分析は不得意である
# 2.単純ベイズ(ナイーブベイズ)
# 長所
# 　　-数式で出来る為、モデルの構造が分かり易い。
# 　　-数式で出来る為、予測が可能である。
# 　　-ベイズの定理を用いて、恣意性を取り入れる事が出来る。
# 短所
# 　　-非線形に比べてモデルの表現力が低い。
# 　　-特徴量を高次(2乗、3乗..)にした際に過学習を起こしやすい。
# 　　-離散型のデ－タ分析は不得意である
# 3.SVM
# 長所
# 　　-数式で出来る為、モデルの構造が分かり易い。
# 　　-数式で出来る為、予測が可能である。
# 　　-生存予測（分類）の他に、回帰でも用いる事が出来る。また、非線形にも対応している。
# 短所
# 　　-非線形に比べてモデルの表現力が低い。
# 　　-特徴量を高次(2乗、3乗..)にした際に過学習を起こしやすい。
# 　　-離散型のデ－タ分析は不得意である
# 4.ランダムフォレスト
# 長所
# 　　-計算速度が速い。
# 　　-どの特徴量が重要か把握出来る。
# 　　-データの正規化・標準化を考える必要がない。
# 短所
# 　　-決定木の深さ(max_depth)を深さを多くすればするほど、過学習し易い。
# 　　-モデルの精度が良かった場合でも、なぜ良かったのか分かり辛い。
# 　　-表現力の高いモデルである分、線形と比べて精度を出す為のデータ量が多くなる。
### モデル選択の基準
# 1.精度
# 2.トレーニング時間
# 3.パラメーターの数
# 4.特徴の数

### モデル選択におけるデータ可視化
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

sns.barplot(x="Age",y="Survived",hue='Sex',data=data_train)
sns.pointplot(x="Pclass", y="Survived", hue="Sex", data=data_train)

### データ可視化の結果について考察せよ
# 上記の２つの可視化から分かったことについて考察せよ。
# 20代以上の年齢層は、男性よりも女性の生存率が高い。
# 30代の女性の生存率が最も高い。
# Pclassが高いほどが生存率が高い。
# いずれのPclassも男性よりも女性のほうが生存率が高い。

# 上記の考察結果から、モデル選択を考える場合、どのようなことが考えられるか
# 今回、生存予測（分類）を行なう上で、生存したか否かを示す特徴量(Survived)と
# 相関の高い特徴量を調べる事で、精度の高い結果を導く事が出来ると想定。
# よって、どの特徴量が重要か把握出来るランダムフォレストを使用する。

### データセットの分割
from sklearn.model_selection import train_test_split
X = data_train.drop(['Survived','PassengerId'],axis=1)
y = data_train['Survived']
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2)

### ランダムフォレストについて記述せよ
#　決定木とはどのような手法か
#  樹木状のモデルを使用した分類を行なう。
#　ランダムフォレストとはどのような手法か
#　 ランダムフォレストは、アンサンブル学習法(複数の分類器を集めて構成される分類器)の一つで、
#  複数の決定木を作り多数決で分類を行う。
#　ランダムフォレストの長所と短所をそれぞれ3つ以上挙げてください。
#  長所
#  　　-計算速度が速い。
#  　　-どの特徴量が重要か把握出来る。
#  　　-データの正規化・標準化を考える必要がない。
#  短所
#  　　-決定木の深さ(max_depth)を深さを多くすればするほど、過学習し易い。
#  　　-モデルの精度が良かった場合でも、なぜ良かったのか分かり辛い。
#  　　-表現力の高いモデルである分、線形と比べて精度を出す為のデータ量が多くなる。
#　今回の目的からランダムフォレストの手法が適する理由を考察し、記述せよ
#   データ量が十分にある為、表現力が高い、非線形のモデルが有効であり、
#   特徴量の重要度が分かり、かつ、学習にかかる計算量が比較的少ない
#   ランダムフォレストの手法が適している。

### 学習
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(n_estimators=12,criterion='gini',max_depth=9,random_state=None)
clf.fit(X_train,y_train)

### accuracyを求めよ
from sklearn.metrics import accuracy_score
predictions =clf.predict(X_test)#
print(accuracy_score(y_test,predictions)) #0.815642458101

### 精度を高める
print(accuracy_score(y_test,predictions)) #0.815642458101

### ハイパーパラメータについて
# ハイパーパラメーターとは何か
# モデルの複雑さを決めるような次数や正則化パラメータの事
# ランダムフォレストにおいてどのようなハイパーパラメーターがあるか4つ以上記述せよ
# n_estimators
# criterion
# max_features
# max_depth
# 記述したハイパーパラメーターにおいて、それぞれどのような値が存在するか記述せよ
# （そのハイパーパラメーターを変化させるとどのようなことが起きるかも記述すること）
# n_estimators:決定木の数を指定する。数を増やすと精度が上がる。ただし、ある時点で数を増やしても精度は上がらなくなる。(default=10)
# criterion:データ分割基準。"gini"と"entropy"があり、学習にかかる時間と精度が変化する。"gini"のほうが学習にかかる時間が少ないとされている。(default=”gini”)
# max_features:特徴量の最大数。int、float、string、None、"Auto","sqrt","log2"がある。
#              値が大きいと決定木が似たようなものになり、最も識別性の高い特徴量をつかうので、訓練データに容易に適合できるので、精度が上がる。(default=”auto”)
# max_depth:決定木の深さの最大値。数を増やすと精度が上がる。ただし、ある時点で数を増やしても精度は上がらなくなる。(default=None)

### チューニング
#パラメーター設定
clf_tune_1 = RandomForestClassifier(n_estimators=4)
clf_tune_2 = RandomForestClassifier(n_estimators=6)
clf_tune_3 = RandomForestClassifier(n_estimators=9)
clf_tune_4 = RandomForestClassifier(max_features='log2')
clf_tune_5 = RandomForestClassifier(max_features='sqrt')
clf_tune_6 = RandomForestClassifier(max_features='auto')
clf_tune_7 = RandomForestClassifier(criterion='entropy')
clf_tune_8 = RandomForestClassifier(criterion='gini')
clf_tune_9 = RandomForestClassifier(max_depth=2)
clf_tune_10 = RandomForestClassifier(max_depth=3)
clf_tune_11 = RandomForestClassifier(max_depth=5)
clf_tune_12 = RandomForestClassifier(max_depth=10)
clf_tune_13 = RandomForestClassifier(min_samples_split=2)
clf_tune_14 = RandomForestClassifier(min_samples_split=3)
clf_tune_15 = RandomForestClassifier(min_samples_split=5)
clf_tune_16 = RandomForestClassifier(min_samples_leaf=1)
clf_tune_17 = RandomForestClassifier(min_samples_leaf=5)
clf_tune_18 = RandomForestClassifier(min_samples_leaf=8)

#各パラメーターで学習
clf_tune_1.fit(X_train,y_train)
clf_tune_2.fit(X_train,y_train)
clf_tune_3.fit(X_train,y_train)
clf_tune_4.fit(X_train,y_train)
clf_tune_5.fit(X_train,y_train)
clf_tune_6.fit(X_train,y_train)
clf_tune_7.fit(X_train,y_train)
clf_tune_8.fit(X_train,y_train)
clf_tune_9.fit(X_train,y_train)
clf_tune_10.fit(X_train,y_train)
clf_tune_11.fit(X_train,y_train)
clf_tune_12.fit(X_train,y_train)
clf_tune_13.fit(X_train,y_train)
clf_tune_14.fit(X_train,y_train)
clf_tune_15.fit(X_train,y_train)
clf_tune_16.fit(X_train,y_train)
clf_tune_17.fit(X_train,y_train)
clf_tune_18.fit(X_train,y_train)

#各パラメーターの学習結果を元に予測
predictions_1 = clf_tune_1.predict(X_test)
predictions_2 = clf_tune_2.predict(X_test)
predictions_3 = clf_tune_3.predict(X_test)
predictions_4 = clf_tune_4.predict(X_test)
predictions_5 = clf_tune_5.predict(X_test)
predictions_6 = clf_tune_6.predict(X_test)
predictions_7 = clf_tune_7.predict(X_test)
predictions_8 = clf_tune_8.predict(X_test)
predictions_9 = clf_tune_9.predict(X_test)
predictions_10 = clf_tune_10.predict(X_test)
predictions_11 = clf_tune_11.predict(X_test)
predictions_12 = clf_tune_12.predict(X_test)
predictions_13 = clf_tune_13.predict(X_test)
predictions_14 = clf_tune_14.predict(X_test)
predictions_15 = clf_tune_15.predict(X_test)
predictions_16 = clf_tune_16.predict(X_test)
predictions_17 = clf_tune_17.predict(X_test)
predictions_18 = clf_tune_18.predict(X_test)

#各パラメーターの精度を表示
print(accuracy_score(y_test,predictions_1))
print(accuracy_score(y_test,predictions_2))
print(accuracy_score(y_test,predictions_3))
print(accuracy_score(y_test,predictions_4))
print(accuracy_score(y_test,predictions_5))
print(accuracy_score(y_test,predictions_6))
print(accuracy_score(y_test,predictions_7))
print(accuracy_score(y_test,predictions_8))
print(accuracy_score(y_test,predictions_9))
print(accuracy_score(y_test,predictions_10))
print(accuracy_score(y_test,predictions_11))
print(accuracy_score(y_test,predictions_12))
print(accuracy_score(y_test,predictions_13))
print(accuracy_score(y_test,predictions_14))
print(accuracy_score(y_test,predictions_15))
print(accuracy_score(y_test,predictions_16))
print(accuracy_score(y_test,predictions_17))
print(accuracy_score(y_test,predictions_18))

### 検証について記述せよ
# 検証は何か
# UnderFittingとOverFittingが起こっている事を確認する。

# なぜ検証を行う必要があるのか(Accuracyだけではダメな理由も含めること)
# 汎用的なモデルになっているか確認する必要がある為で、Accuracyだけでは過学習が起きている恐れがある。
# 主な検証方法について2つ以上記述せよ
# ホールドアウト法とK分割交差検証法
### KFoldについて記述せよ
# K-分割交差検証について説明せよ
# データをK個に分割し、そのうちの1つをテストとし、残る K−1 個のデータを学習用データとする。
# K-分割交差検証はデータセットを何個に分割するか
# 指定したK個に分割する。
# データセットを分割する際、その個数はどのように考えると良いか
# K−1個のデータが学習用データとなり、残る1つがテスト用データとなる。
# K-分割交差検証は何回の検証を行うか
# 指定したK回の検証を行なう。
# K-分割交差検証の結果は、最終的にどのように求められるか
# それぞれの回で導き出した精度を平均化し、求める。

### KFoldを実施する
from sklearn.model_selection import cross_val_score
n_fold = 10 #交差検定の回数
scores = cross_val_score(clf,X,y,cv=n_fold)
### Kfoldの結果を確認する
print(scores)
print(scores.mean()) #0.831631199637

### どの特徴量が重要であったかを調査する
print(clf.feature_importances_)
#[ 0.08372367  0.24143289  0.07578829  0.05694132  0.04086564  0.07443233
#  0.05319167  0.18180694  0.19181725]
print(X)
#-> Xと紐付けると、Sexが最も重要で、次にNamePrefix、Lnameが重要であった事が分かる。
#   X = data_train.drop(['Survived','PassengerId'],axis=1)
